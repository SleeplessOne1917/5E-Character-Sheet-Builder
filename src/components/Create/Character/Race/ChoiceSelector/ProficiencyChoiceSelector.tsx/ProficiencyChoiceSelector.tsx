'use client';

import {
	SrdProficiencyItem,
	SrdProficiencyItemChoice
} from '../../../../../../types/srd';
import {
	addProficiency,
	removeProficiency
} from '../../../../../../redux/features/proficiencies';
import {
	addTraitProficiency,
	removeTraitProficiency
} from '../../../../../../redux/features/raceInfo';
import {
	useAppDispatch,
	useAppSelector
} from '../../../../../../hooks/reduxHooks';
import { useCallback, useEffect, useState } from 'react';

import ChoiceSelector from '../ChoiceSelector';
import Select from '../../../../../Select/Select/Select';

type OptionSelectorProps = {
	choice: SrdProficiencyItemChoice;
	traitIndex: string;
	label: string;
};

const ProficiencyChoiceSelector = ({
	choice,
	traitIndex,
	label
}: OptionSelectorProps): JSX.Element => {
	const dispatch = useAppDispatch();
	const selectedProficiencies = useAppSelector(
		state =>
			state.editingCharacter.raceInfo.selectedTraitProficiencies[traitIndex]
	);

	const getInitialSelectValues = useCallback(() => {
		const returnValues: string[] = [];

		for (let i = 0; i < choice.choose; ++i) {
			returnValues.push('blank');
		}

		return returnValues;
	}, [choice]);

	const [selectValues, setSelectValues] = useState<string[]>(
		selectedProficiencies?.map(({ index }) => index) ?? getInitialSelectValues()
	);

	const proficiencies = useAppSelector(
		state => state.editingCharacter.proficiencies
	);

	useEffect(() => {
		if (!selectedProficiencies || selectedProficiencies.length === 0) {
			setSelectValues(getInitialSelectValues());
		}
	}, [setSelectValues, getInitialSelectValues, selectedProficiencies]);

	const handleChangeSelect = useCallback(
		(index: number, selectValue: string) => {
			if (selectValues[index] !== 'blank') {
				dispatch(
					removeTraitProficiency({
						index: traitIndex,
						proficiency: selectValues[index]
					})
				);
				dispatch(removeProficiency(selectValues[index]));
			}

			if (selectValue !== 'blank') {
				const proficiency = (
					choice.from.options.map(option => option.item) as SrdProficiencyItem[]
				).find(prof => prof.index === selectValue) as SrdProficiencyItem;

				dispatch(addTraitProficiency({ index: traitIndex, proficiency }));
				dispatch(addProficiency(proficiency));
			}

			setSelectValues(prev =>
				prev.map((value, i) => (i === index ? selectValue : value))
			);
		},
		[dispatch, selectValues, traitIndex, choice.from.options, setSelectValues]
	);

	const handleReset = useCallback(() => {
		for (const { index } of choice.from.options
			.filter(option => option.item && selectValues.includes(option.item.index))
			.map(option => option.item) as SrdProficiencyItem[]) {
			dispatch(
				removeTraitProficiency({ index: traitIndex, proficiency: index })
			);
			dispatch(removeProficiency(index));
		}

		setSelectValues(getInitialSelectValues());
	}, [
		choice,
		selectValues,
		dispatch,
		traitIndex,
		setSelectValues,
		getInitialSelectValues
	]);

	const selects: JSX.Element[] = [];

	for (let i = 0; i < choice.choose; ++i) {
		selects.push(
			<Select
				key={i}
				value={selectValues[i]}
				onChange={value => handleChangeSelect(i, value as string)}
				options={[{ value: 'blank', label: '\u2014' }].concat(
					choice.from.options
						.filter(
							option =>
								option.item &&
								(!(
									selectValues.includes(option.item.index) ||
									proficiencies
										.map(proficiency => proficiency.index)
										.includes(option.item.index)
								) ||
									option.item.index === selectValues[i])
						)
						.map(option => ({
							value: option.item?.index as string,
							label: option.item?.name.replace(/Skill: /g, '') as string
						}))
				)}
			/>
		);
	}

	return (
		<ChoiceSelector
			label={label}
			selects={selects}
			isSelected={!selectValues.includes('blank')}
			onReset={handleReset}
		/>
	);
};

export default ProficiencyChoiceSelector;
