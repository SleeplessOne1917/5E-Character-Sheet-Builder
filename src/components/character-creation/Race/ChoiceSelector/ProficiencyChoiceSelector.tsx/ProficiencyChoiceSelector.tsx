import {
	SrdProficiencyItem,
	SrdProficiencyItemChoice
} from '../../../../../types/srd';
import { useCallback, useState } from 'react';

import ChoiceSelector from '../ChoiceSelector';
import Select from '../../../../Select/Select';
import { useAppSelector } from '../../../../../hooks/reduxHooks';

type OptionSelectorProps = {
	choice: SrdProficiencyItemChoice;
	onReset?: (items: SrdProficiencyItem[]) => void;
	onApply?: (items: SrdProficiencyItem[]) => void;
	initialValues?: string[];
	label: string;
};

const ProficiencyChoiceSelector = ({
	choice,
	onReset = () => {},
	onApply = () => {},
	initialValues,
	label
}: OptionSelectorProps): JSX.Element => {
	const getInitialSelectValues = useCallback(() => {
		const returnValues: string[] = [];

		for (let i = 0; i < choice.choose; ++i) {
			returnValues.push('blank');
		}

		return returnValues;
	}, [choice]);

	const [selectValues, setSelectValues] = useState<string[]>(
		initialValues ?? getInitialSelectValues()
	);

	const proficiencies = useAppSelector(
		state => state.editingCharacter.proficiencies
	);

	const handleChangeSelect = useCallback(
		(index: number, selectValue: string) =>
			setSelectValues(prevState =>
				prevState.map((value, i) => (i === index ? selectValue : value))
			),
		[setSelectValues]
	);

	const handleApply = useCallback(
		() =>
			onApply(
				choice.from.options
					.filter(option => selectValues.includes(option.item.index))
					.map(option => option.item) as SrdProficiencyItem[]
			),
		[onApply, choice, selectValues]
	);

	const handleReset = useCallback(() => {
		setSelectValues(getInitialSelectValues());
		onReset(
			choice.from.options
				.filter(option => selectValues.includes(option.item?.index))
				.map(option => option.item) as SrdProficiencyItem[]
		);
	}, [setSelectValues, getInitialSelectValues, onReset, choice, selectValues]);

	const selects: JSX.Element[] = [];

	for (let i = 0; i < choice.choose; ++i) {
		selects.push(
			<Select
				key={i}
				value={selectValues[i]}
				onChange={value => handleChangeSelect(i, value as string)}
				options={[{ value: 'blank', label: '\u2014' }].concat(
					choice.from.options
						.filter(
							option =>
								option.item &&
								(!(
									selectValues.includes(option.item.index) ||
									proficiencies
										.map(proficiency => proficiency.index)
										.includes(option.item.index)
								) ||
									option.item.index === selectValues[i])
						)
						.map(option => ({
							value: option.item?.index as string,
							label: option.item?.name.replace(/Skill: /g, '') as string
						}))
				)}
			/>
		);
	}

	return (
		<ChoiceSelector
			label={label}
			selectValues={selectValues}
			selects={selects}
			isSelected={!!initialValues}
			onApply={handleApply}
			onReset={handleReset}
		/>
	);
};

export default ProficiencyChoiceSelector;
