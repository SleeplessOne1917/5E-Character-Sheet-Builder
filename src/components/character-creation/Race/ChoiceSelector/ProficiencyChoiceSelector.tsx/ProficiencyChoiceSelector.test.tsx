import * as stories from './ProficiencyChoiceSelector.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/testing-react';

const { Default } = composeStories(stories);

it('renders correctly', () => {
	render(<Default />);

	expect(screen.getByTestId('choice-selector')).toMatchSnapshot();
});
