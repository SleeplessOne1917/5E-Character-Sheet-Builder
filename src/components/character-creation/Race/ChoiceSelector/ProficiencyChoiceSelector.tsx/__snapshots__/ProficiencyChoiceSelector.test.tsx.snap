// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="selector"
  data-testid="choice-selector"
>
  <div
    class="label"
  >
    Choose stuff
  </div>
  <div
    class="select-div"
  >
    <select
      aria-label="Select choice"
      class="select"
    >
      <option
        value="blank"
      >
        —
      </option>
      <option
        value="foo"
      >
        Foo
      </option>
      <option
        value="bar"
      >
        Bar
      </option>
      <option
        value="baz"
      >
        Baz
      </option>
    </select>
    <select
      aria-label="Select choice"
      class="select"
    >
      <option
        value="blank"
      >
        —
      </option>
      <option
        value="foo"
      >
        Foo
      </option>
      <option
        value="bar"
      >
        Bar
      </option>
      <option
        value="baz"
      >
        Baz
      </option>
    </select>
  </div>
  <div
    class="button-div"
  >
    <button
      class="button negative"
      style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
      type="button"
    >
      Reset
    </button>
    <button
      class="button positive"
      disabled=""
      style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
      type="button"
    >
      Apply
    </button>
  </div>
</div>
`;
